### PR TITLE
KAFKA-6253: Improve sink connector topic regex validation

### PR DESCRIPTION
**Pull request by jklukas**
KAFKA-3073 added topic regex support for sink connectors. The addition requires that you only specify one of topics or topics.regex settings. This is being validated in one place, but not during submission of connectors. This PR adds validation at `AbstractHerder.validateConnectorConfig` and `WorkerConnector.initialize`.

This adds a test of the new behavior to `AbstractHerderTest`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (jklukas) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: I changed the logic for determining source or sink due to issues in testing. In the herder tests, mocked connectors show a class of `org.apache.kafka.connect.connector.Connector$$EnhancerByCGLIB$$d27185f2`, which is neither a SinkConnector nor a SourceConnector, so the existing `instanceof` check always returns false in the tests.

I'd be happy to revert this logic change if the `Class.forName(connType)` is undesirable, but I'd need help understanding why the mocks appear as `Connector` rather than the subtype that appears in the configs.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java
Comment (jklukas) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java:38]: Making this member public to be symmetric with TOPICS_CONFIG

**Comment by jklukas** [comment]
cc @rhauch @ewencp

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java]: Minor: should be "Throw an exception *if* ..."

**Review by rhauch — APPROVED** [review]
One typo in the JavaDoc, but once that's fixed then we can merge this. Thanks, @jklukas!

**Comment by jklukas** [comment]
Thanks for the review, @rhauch. Typo fixed.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java
Comment (ewencp) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java]: There might be a few cases where we used `Connector` when creating the mock if it didn't previously matter to the test. e.g. https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/DistributedHerderTest.java#L355 is one case I can find. https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java#L635 might be another case, where we used Connector since I think both source and sink tests might use that same code.

I guess the `Class.forName` is probably ok since we've swapped in the correct `ClassLoader` at this point, though not really ideal as we're doing more lookups when we already have the instance of the connector. I think it would be better to fix the tests (though I'm not sure off the top of my head how many changes this will require...)

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:80]: Hmm, does validation here do much? Since we're not validating a bunch of other stuff here (i.e. all the validation of the base ConnectorConfig settings, anything defined by the connector itself, etc) there's still a lot of ways we could fail.

If we validate input before ever writing it to the config topic, it seems like we should catch most issues. I think having validation on starting a connector could potentially allow for better error messages, so may be worth doing. But could we just reuse the validation in AbstractHerder instead? For example, I think this method is only called by `Worker.startConnector`, which in turn is only called in two places by StandaloneHerder and DistributedHerder. Could we just add extra calls to `AbstractHerder.validateConnectorConfig` at those locations before we try to invoke this code?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: not critical since it's not a ton of logic, but since this logic is repeated, it might be better to turn it into a utility method on `SinkConnectorConfig` and use it both in that class's validate method and here.

**Review by ewencp — COMMENTED** [review]
@jklukas thanks for the follow up, and sorry for the delay in reviewing it. this mostly looks good, i just had a few minor comments. i left a note about the class being used by mocks, let me know if you want any further help looking into how to make them work with that conditional

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (jklukas) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:80]: You're correct about where this method gets called within the main code, but there are repercussions to making an additional call to `AbstractHerder.validateConnectorConfig ` for the mocks used in tests. I can investigate this, but it will take more time.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java
Comment (jklukas) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/standalone/StandaloneHerderTest.java:638]: This signature without an explicit connector mock was a source of most of the failing tests due to `Connector.class` being used for the default mock, as @ewecp suggested. I'm removing this variant in favor of always explicitly passing a connectorMock that uses the appropriate subclass.

**Comment by jklukas** [comment]
I've pushed new commits that address @ewecp's two major concerns. I haven't made the change of pushing validation out of `WorkerConnecter::initialize`. I'll try to get back to making that change in the next week or two if @ewencp thinks it's worth waiting to get that in.

**Comment by rhauch** [comment]
@jklukas, any more process? Or do you think that this is okay as-is?

**Comment by jklukas** [comment]
I'm comfortable with this as-is. I haven't fully wrapped my head around what benefit we'd get from doing the last refactoring that @ewencp suggested, but I defer to him as to whether he's comfortable moving ahead without it.

**Comment by ewencp** [comment]
Going to close and reopen to get tests to run again. The failures are unrelated on JDK7 and 8, and on JDK9 the failure is `WorkerSourceTask.run(): expected: 1, actual: 1`, which could be a threading issue or some other brokenness.

**Comment by jklukas** [comment]
Squashed and rebased to let tests run again.

**Comment by jklukas** [comment]
Looks like the JDK 8 tests failed with `15:08:29 :core:testScoverageBuild timed out (after 240 minutes). Marking the build as aborted.`

**Review by ewencp — APPROVED** [review]
LGTM, thanks for the follow up!

**Pull request closed without merging**
